converter/converter.go

```diff
@@ -0,0 +1,437 @@
+	// The converted proto to gunk declarations. This only stores
+	// the messages, enums and services. These get converted to as
+	// they are found.
+	translatedDeclarations []*strings.Builder
+
+	// If an error is found during processing store the error
+	// and error position here.
+	err    error
+	errPos scanner.Position
+
+	// The package, option and imports from the proto file.
+	// These are converted to gunk after we have converted
+	// the rest of the proto declarations.
+	pkg     *proto.Package
+	pkgOpts []pkgOpt
+	imports []*proto.Import
+}
+
+// Convert will convert a proto file to a gunk file,
+// saving the file in the same folder as the proto file.
+func Convert(path string, overwrite bool) error {
+	if filepath.Ext(path) != ".proto" {
+		return fmt.Errorf("convert requires a .proto file")
+	}
+	reader, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("unable to read file %q: %v", path, err)
+	}
+	defer reader.Close()
+
+	// Parse the proto file.
+	parser := proto.NewParser(reader)
+	d, err := parser.Parse()
+	if err != nil {
+		return fmt.Errorf("unable to parse proto file %q: %v", path, err)
+	}
+
+	// Start converting the proto declarations to gunk.
+	b := builder{}
+	for _, e := range d.Elements {
+		if err := b.handleProtoType(e); err != nil {
+			return fmt.Errorf("error: %v\n", err)
+		}
+		if b.err != nil {
+			return fmt.Errorf("error in %d:%d: %v\n", b.errPos.Line, b.errPos.Column, b.err)
+		}
+	}
+
+	// Convert the proto package and imports to gunk.
+	translatedPkg := b.handlePackage()
+	translatedImports := b.handleImports()
+
+	// Add the converted package and imports, and then
+	// add all the rest of the converted types. This will
+	// keep the order that things were declared.
+	w := &strings.Builder{}
+	w.WriteString(translatedPkg.String())
+	w.WriteString("\n\n")
+	w.WriteString(translatedImports.String())
+	w.WriteString("\n")
+	for _, l := range b.translatedDeclarations {
+		w.WriteString("\n")
+		w.WriteString(l.String())
+		w.WriteString("\n")
+	}
+
+	// TODO: We should run this through the Gunk generator to
+	// make sure that it compiles?
+	// TODO: Run this through the gunk fmt when it is implemented.
+
+	fileToWrite := strings.Replace(filepath.Base(path), ".proto", ".gunk", 1)
+	fullpath := filepath.Dir(path) + "/" + fileToWrite
+
+	if _, err := os.Stat(fullpath); !os.IsNotExist(err) && !overwrite {
+		return fmt.Errorf("path already exists %q, cannot overwrite", fullpath)
+	}
+	if err := ioutil.WriteFile(fullpath, []byte(w.String()), 0644); err != nil {
+		return fmt.Errorf("unable to write to file %q: %v", fullpath, err)
+	}
+
+	return nil
+}
+
+// Write will write output to a string builder, adding in indentation
+// where required. It will write the comment first if there is one,
+// and then write the rest.
+func (b *builder) Write(w *strings.Builder, indent int, comments *proto.Comment, s string, args ...interface{}) {
+	if comments != nil {
+		for _, c := range comments.Lines {
```

Comment:
does this deal with inline comments, e.g.

```
some proto line // inline comment
```

I presume not, in which case we need to either error in those cases, or add a TODO to fix them

Reply:
Ah, missed that, good catch! I have left a TODO. It might be a bit "hacky" to do, so I will deal with it later.